libbeat/statestore/backend/es/notifier.go

```diff
@@ -0,0 +1,66 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package es
+
+import (
+	"sync"
+
+	conf "github.com/elastic/elastic-agent-libs/config"
+)
+
+type OnConfigUpdateFunc func(c *conf.C)
+
+type Notifier struct {
+	mx sync.RWMutex
+
+	listeners map[int]OnConfigUpdateFunc
+	counter   int
+}
+
+func NewNotifier() *Notifier {
+	n := &Notifier{
+		listeners: make(map[int]OnConfigUpdateFunc),
+	}
+	return n
+}
+
+func (n *Notifier) Subscribe(fn OnConfigUpdateFunc) int {
+	n.mx.Lock()
+	defer n.mx.Unlock()
+
+	id := n.counter
+	n.counter++
+	n.listeners[id] = fn
+
+	return id
+}
+
+func (n *Notifier) Unsubscribe(id int) {
+	n.mx.Lock()
+	defer n.mx.Unlock()
+	delete(n.listeners, id)
```

Comment:
Why does this decrement n.counter?

Reply:
counter here is just the id (handle) that is returned to the user on Subscribe in order to allow to that "handle" later for Unsubscribe.